### ISSUE TITLE
type_id is not sufficiently collision-resistant

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#129014]
The soundness of functions like [`downcast`](https://doc.rust-lang.org/nightly/std/boxed/struct.Box.html#method.downcast-2) relies on the type_id of two different types never being equal. Currently, the type_id is a 128-bit hash of the full type identity, computed specifically via SipHash-1-3 with an all-zero key. This is not a strong enough hash function for this purpose.

The lang team [decided](https://github.com/rust-lang/rust/pull/95845#issuecomment-1175324596) that relying on a full (non-truncated) cryptographic hash is fine -- we don't have to guarantee soundness against an infinite-resource attacker that can generate collisions in cryptographic hash functions.
However, SipHash even in its default configuration (SipHash-2-4) is *not* a cryptographic hash, as clarified [by its author](https://github.com/rust-lang/rust/issues/10389#issuecomment-1887239511)[^siphash] -- it is a pseudo-random function (PRF); that means it assumes a secret key, but Rust hard-codes an all-zero key and in fact has no way to keep a key secret. By standards for cryptographic hash functions, SipHash-2-4 with an all-zero-key is [weaker than MD5](https://github.com/rust-lang/rust/issues/10389#issuecomment-2181169783), and generating a collision for that [is pretty easy](https://x.com/realhashbreaker/status/1770161965006008570) these days. SipHash-1-3 is even weaker, we should thus expect it to be a matter of hours to create a collision, if someone really tried.

[^siphash]: In the [original paper](https://cr.yp.to/siphash/siphash-20120918.pdf), they even write: "We comment that SipHash is not meant to be, and (obviously) is not, collision-resistant."

Generating a concrete example of an unsoundness from that is a bit more tricky since one would have to find a Rust type generating the collision, but it seems fairly clear that the bar of "full (non-truncated) cryptographic hash" is *not* met by the current type_id implementation. The point of this issue is to determine how the compiler implementation can best satisfy the soundness expectation set by the lang team.

If you instead want to argue that the lang team should change its mind, please open a new issue and gather arguments in favor of that position, so that a summary of all the arguments for either option can be brought to the lang team for discussion. Also, this issue is only about type_id. According to [this comment](https://github.com/rust-lang/rust/issues/10389#issuecomment-2284095948), the only other case where a has collision could cause unsoundness is with incremental builds. All other hashes are actively checked for collisions by Rust. That said, it is not clear to me whether that also covers dynamic linking scenarios. Either way, if you are concerned about soundness issues caused by hash collisions outside of type_id, please file a separate issue.

### Possible solutions

We should do one of the following:
1. switch to a stronger hash function, or
2. switch to a different scheme that doesn't rely on collision-resistance of the hash function.

Worth noting is that a cryptographic hash function these days must output [at least 256 bits](https://github.com/rust-lang/rust/issues/10389#issuecomment-2243114649) to be considered worth its salt. The lang team has not spelled out their exact definition of "cryptographic hash function", but the [standard definition](https://en.wikipedia.org/wiki/Cryptographic_hash_function) includes collision resistance, and in fact collisions are exactly what we are most worried about here, so it seems reasonable to assume that this is part of the lang team intent. A lang team member [mentioned](https://github.com/rust-lang/rust/pull/95845#issuecomment-1169054470) BLAKE3 as a candidate, further corroborating this claim. So we'd have to switch to BLAKE3 or SHA2 or something like that.

Therefore, in both cases will we need more than the currently available 128 bits of a `TypeId` to obtain the desired level of collision resistance. To avoid further increasing the size of `TypeId`, the most likely scheme would be to include a pointer in `TypeId` that points to the remaining information -- either a 256-bit hash, or a string (null-terminated or with leading length information, to obtain a "thin" pointer), or something like that.

https://github.com/rust-lang/rust/pull/95845 explores what this could look like without depending on a hash function: `TypeId` becomes a pair of a (not-soundness-critical) hash and a pointer. If the hashes are different, we can quickly conclude "inequal". If the pointers are equal, we can quickly conclude "equal". (Many linkers should be able to deduplicate the data the pointers point to, making the "equal" optimization even work cross-crate in many cases.) The same approach could easily be used without a full type name, by storing a pair of a low-quality hash for quick "inequal" checks and a high-equality hash behind the pointer to provide soundness.

*Relying* on the linker to deduplicate is unlikely to work since not all platforms have linkers that can do that (in particular for dynamic linking). C++ has a similar problem to solve and, at least on Windows, [seems to](https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/Const.20.60TypeId.3A.3Aof.60.2C.20.60TypeId.3A.3Amatches.60.2C.20.60type_name.60/near/450558553) do something like the first of these options: compare pointers, and fall back to comparing strings.

Therefore, the next step here seems to be for the compiler team to decide whether to use a hash function or to include the type name in the binary. The latter has an existing [implementation](https://github.com/rust-lang/rust/pull/95845), but there were concerns about leaking type names (and the paths leading to those types) in the binary.

**Comment by the8472** [comment]
> If you instead want to argue that the lang team should change its mind, please open a new issue and gather arguments in favor of that position

I have opened #129030 which somewhat goes in that direction, though from a different angle.

**Comment by apiraino** [comment]
Discussed in T-compiler triage meeting [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202024-08-15/near/462568536). The topic needs more time, is worth a dedicated design meeting. Not only about the options we have but also about defining the threat model we are trying to protect against.

**Comment by briansmith** [comment]
> All other hashes are actively checked for collisions by Rust.

When a collision is detected, compilation will fail, right? If so, the issue is reduced from unsoundness to denial of service; that's definitely a less-serious issue but still problematic.

**Comment by RalfJung** [comment]
This issue is about type_id; if you are concerned about other hashes please open a new issue or discuss that in https://github.com/rust-lang/rust/issues/129030.

**Comment by briansmith** [comment]
> So we'd have to switch to BLAKE3 or SHA2 or something like that.

[The discussion from the Bazel team regarding switching from SHA-256 to BLAKE3](https://github.com/bazelbuild/bazel/discussions/22011) maybe be useful. On modern x86-64 and ARMv8 hardware the fastest SHA-256 implementations that use SHA-256-specific CPU instructions should be notably faster than BLAKE3.

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/channel/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23129014.20type_id.20is.20not.20sufficiently.20collision-resistant)).

@rustbot label -I-prioritize +C-discussion

**Comment by RalfJung** [comment]
That sounds to me like an implementation of that scheme (basically a revival of https://github.com/rust-lang/rust/pull/95845) should be accepted? Or would an MCP still be required?

The main remaining concern seems to be @the8472 who is [concerned](https://github.com/rust-lang/rust/issues/129030#issuecomment-2619411566) about inconsistent decisions in this space.  However, I don't see why that would block a PR if the change can be demonstrated to have negligible downsides.

**Comment by the8472** [comment]
Yeah, as I said in https://github.com/rust-lang/rust/issues/129030#issuecomment-2285883346 

> AIUI one of the proposed fixes is storing all the type names in a binary. This will increase binary sizes which is a real issue for embedded systems. If there's a fix that's free in terms of the produced output then I agree that it's not productive to argue further.

at that point the threat model discussion would become only about general policy, not the typeid issue.

**Comment by bjoernager** [comment]
If the memory concerns are problematic enough, would adding a compiler flag (e.g. `unsafe-type-id`) that always makes use of hashes similar to the current implementation be an appropriate mitigation?

<sub>Assuming the conditional layout/representation of the type from such a flag wouldn't pose any issues.</sub>

**Comment by fu5ha** [comment]
Hi, I'm interested in working on implementation for this, be it a revival of https://github.com/rust-lang/rust/pull/95845, something similar but storing a 256-bit hash behind the pointer rather than the full mangled string, or simply increasing the efficacy of the hashing algorithm while leaving the result as a 128-bit hash inline. However, it seems ill advised to begin on implementing any of these without agreement on what the threat model is, and therefore using that to inform which option is the best to move forward with (thus hopefully preventing the implementation from again being stuck in limbo due to lack of consensus on the intended goals of the change). For example, if we decide that we don't care about protecting against *engineered* collisions, only against "legitimate" ones, it could change our outlook on what design we could accept for a hash function.

Would it be possible to add that conversation to the docket at some point in the near future?

> If the memory concerns are problematic enough, would adding a compiler flag (e.g. unsafe-type-id) that always makes use of hashes similar to the current implementation be an appropriate mitigation?

This seems like a poor compromise to me, and also requires writing, maintaining, and ensuring two different paths function as expected. If the memory concerns are problematic enough, we should probably try to find an alternate solution.

**Comment by the8472** [comment]
> However, it seems ill advised to begin on implementing any of these without agreement on what the threat model is

That's #129030 and it's currently waiting on t-lang.

**Comment by fu5ha** [comment]
Sorry, yes, perhaps I should have posted there but the idea of my post was more or less to inquire about t-lang having the necessary discussion as you say.

**Comment by RalfJung** [comment]
This also recently came up on a [const-eval Zulip discussion](https://rust-lang.zulipchat.com/#narrow/channel/146212-t-compiler.2Fconst-eval/topic/.23.5Bfeature.28const_type_id.29.5D): supporting type_id in const would be a lot simpler if we made `TypeId` a pointer to some opaque blob of data that we can then enforce is never actually accessed in const-eval and Miri. That data can then easily be a large hash or whatever, it's also easy to change later.

This does come at a cost though, since ensuring two type_id are different requires loading through that pointer.
